File: src/main/java/org/cubexmc/railway/service/BlockSectionManager.java
package org.cubexmc.railway.service;

import java.util.HashSet;
import java.util.Set;

/**
 * Simple skeleton for section occupancy. Real implementation will 
 * map (lineId, fromStopId, toStopId, direction) -> occupiedByTrainId
 */
public class BlockSectionManager {

    private final Set<String> occupiedKeys = new HashSet<>();

    public boolean tryEnter(String sectionKey) {
        return occupiedKeys.add(sectionKey);
    }

    public void leave(String sectionKey) {
        occupiedKeys.remove(sectionKey);
    }

    public boolean isOccupied(String sectionKey) {
        return occupiedKeys.contains(sectionKey);
    }
}

File: src/main/java/org/cubexmc/metro/service/BlockSectionManager.kt
package org.cubexmc.metro.service

/**
 * Simple skeleton for section occupancy. Real implementation will
 * map (lineId, fromStopId, toStopId, direction) -> occupiedByTrainId
 */
class BlockSectionManager {

    private val occupiedKeys: MutableSet<String> = HashSet()

    fun tryEnter(sectionKey: String): Boolean = occupiedKeys.add(sectionKey)

    fun leave(sectionKey: String) {
        occupiedKeys.remove(sectionKey)
    }

    fun isOccupied(sectionKey: String): Boolean = occupiedKeys.contains(sectionKey)
}
